### DISCUSSION TITLE
CSS loaded with dynamic `import()` is leaky

### DISCUSSION THREAD

**Discussion by MichaelRoosz** [vikejs/vike#1324] [Discussion]
We use Vike with Vue3 and SSR.
Components are loaded async with Vue's "defineAsyncComponent()".

- After starting the server the CSS assets added by SSR are correct for the first request
- For the next request, both the current as well as the css assets from the previous request get added by Vike' SSR
- This continues until for every request all css assets of all components get added

This makes loading the components async pointless, since after every component has been loaded once by any request all future requests will always get all assets

How to fix this?

**Comment by lourot** [comment]
Hi @MichaelRoosz, thanks for raising this. I haven't seen this before. Can you push a minimal reproducer somewhere? Thanks!

**Comment by MichaelRoosz** [comment]
Hello @AurelienLourot ,

thank you for your response.

Here is a minimal example based on "npm init vike": https://github.com/MichaelRoosz/vike-app

The changes made you may find here: https://github.com/MichaelRoosz/vike-app/commit/23a3eadd4dd7193ea9589747e0e5f8605e0f893c

**Reply by brillout** [reply]
When you use [Client Routing](https://vike.dev/client-routing), then the CSS of your previous page aren't removed (since the DOM is preserved upon navigation). Add class names to make sure your CSS applies to what you want.

**Reply by MichaelRoosz** [reply]
I am not using Client Routing, but server side routing. the issue is within the html returned by ssr, not on the client/browser. I will check if the example uses client routing and update to server routing if needed

**Reply by MichaelRoosz** [reply]
also, the problem persists when opening a page in a new browser window or even a different browser on another computer

**Reply by brillout** [reply]
Indeed, I'm just seeing the reproduction is using Server Routing. Also, it's using Vue's SFC so CSS should be scoped anyways. I'm inclined to think it's a Vue issue, not a Vike one.

Try to reproduce with https://github.com/bluwy/create-vite-extra in order to make sure it is indeed not related to Vike. Then report the minimal reproduction to the Vue folks.

**Reply by MichaelRoosz** [reply]
I tried to reproduce it with this template (vue + typescript) - but it seems to work correctly.

Example project: https://github.com/MichaelRoosz/vite-project/
My changes: https://github.com/MichaelRoosz/vite-project/commit/5ed5365162f2a1f67c64b08101bf6c5baeb7c2de

to debug you may visit:
http://localhost:5173/1 (loads component Test1)
http://localhost:5173/2 (loads componen Test2)

**Reply by MichaelRoosz** [reply]
After some debugging, I think it is related to Vike's "retrieveAssetsDev()" / "retrieveAssetsProd()" - the "create-vite-extra" code does not collect css assets in this way as far as I understand, because it is only loading the client entry js code directly within html which then loads css - but Vike is collecting the css assets in order to directly load them via style tags in html.

**Reply by brillout** [reply]
Let me have a look at it tomorrow.

**Reply by MichaelRoosz** [reply]
Here is another example which does not use global components, but has the same problem:
https://github.com/MichaelRoosz/vike-app/tree/example2
changes: https://github.com/MichaelRoosz/vike-app/compare/master...example2

update: I simplified the example a bit more : https://github.com/MichaelRoosz/vike-app/commit/f67079da64a59bb3fb686a100680430b12e9ed1e

I did also test with Nuxt - it has the same problem, and it looks like this nuxt module is able to handle it correctly: https://www.npmjs.com/package/@blokwise/dynamic , but it seems not to be compatible with nuxt 3 (was made for an older nuxt version)

**Reply by brillout** [reply]
> "retrieveAssetsDev()" / "retrieveAssetsProd()" 

Indeed, manually removing the dynamic imports from there solves the issue. I guess Vike shouldn't include dynamic imports, let me fix this.

**Reply by brillout** [reply]
Hm, there doesn't seem to be a way to distinguish a normal import from a dynamic import in Vite's module graph. Maybe you'll have more luck finding a way to tell the difference: see the `collectCss()` method called by `retrieveAssetsDev()`.

**Reply by brillout** [reply]
I can have another look at it sometime next week. (Feel free to ping me if I forget.)

> Maybe you'll have more luck finding a way to tell the difference: see the `collectCss()` method called by `retrieveAssetsDev()`.

It would be a huge help, let me know if you find something.

**Reply by MichaelRoosz** [reply]
After some more debugging, I think we should still add css assets from dynamic imports (not adding them would be a big problem in my setup), but we should check if they belong to a component that has been loaded for the current vue.js app - I think this should be possible, but I am still working on a proof of concept. Once I have something working I will open a pull request.

**Reply by brillout** [reply]
@MichaelRoosz Neat, looking forward to it 👀  (on my side my main blocker was that I couldn't find a way to distinguish a normal import from a dynamic import in Vite's module graph). Keep me udpated how things go (even if it doesn't result in a PR).

**Comment by josetr** [comment]
I'm experiencing the same problem using React 

If I `import "globals1.css"` on `page1/+Page.tsx`, and then navigate using client routing to `page2/+Page.tsx` that imports  `"globals2.css"`, the css from `globals1.css` will still affect page 2.

Also, just hovering over the page 2 link will prefetch it's global css, affecting the current page 1.

Also, while on dev mode, once a css asset has been loaded, it will remain there until I restart the server. F5 isn't enough.

**Reply by brillout** [reply]
See https://vike.dev/faq#why-is-css-leaked-to-other-pages.

> the same problem

OP's issue is around CSS loaded over dynamic `import()`.

**Reply by josetr** [reply]
I see. I thought Vike would allow something similar to https://nextjs.org/docs/app/building-your-application/styling/css-modules#global-styles, where global styles can be imported into any layout, page, or component inside the app directory, unlike pages router, where global styles were only allowed in a single place: _app.tsx.

``` 
"Why is CSS leaked to other pages?"
In a nutshell: CSS is leaky.
```

Doesn't sound right.

`In a nutshell: this is the behavior that we decided to have` is probably more accurate.

Vike knows which CSS files to inject to a page, so I don't see a reason why it should be complicated to remove them before navigating to another page.

FOUC can be a bit annoying, but if it's necessary, it's better than mixing global styles.

**Reply by brillout** [reply]
It isn't that simple. Anyways, I don't see the current situation to be a blocker for users.

**Reply by brillout** [reply]
> Doesn't sound right.

0f2e5f088fec4f6381ab97e0d3c43022f31e8740

**Reply by brillout** [reply]
I further improved https://vike.dev/faq#why-is-css-leaked-to-other-pages. It should clearer.
